File: app/src/main/java/org/autojs/autojs/ui/doc/DocsFragment.java
package org.autojs.autojs.ui.doc;

import android.app.Activity;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.stardust.util.BackPressedHandler;

import org.autojs.autojs.Pref;
import org.autojs.autojs.databinding.FragmentOnlineDocsBinding;
import org.autojs.autojs.ui.main.QueryEvent;
import org.autojs.autojs.ui.main.ViewPagerFragment;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

public class DocsFragment extends ViewPagerFragment implements BackPressedHandler {

    public static final String ARGUMENT_URL = "url";

    private FragmentOnlineDocsBinding binding;
    private WebView mWebView;
    private String mIndexUrl;
    private String mPreviousQuery;

    public DocsFragment() {
        super(ROTATION_GONE);
    }

    public static DocsFragment newInstance(String url) {
        Bundle args = new Bundle();
        args.putString(ARGUMENT_URL, url);
        DocsFragment fragment = new DocsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EventBus.getDefault().register(this);
        setArguments(getArguments() != null ? getArguments() : new Bundle());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentOnlineDocsBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mWebView = binding.ewebView.getWebView();
        SwipeRefreshLayout swipeRefresh = binding.ewebView.getSwipeRefreshLayout();
        swipeRefresh.setOnRefreshListener(this::handleRefresh);
        restoreWebViewState(savedInstanceState);
    }

    private void handleRefresh() {
        if (TextUtils.equals(mWebView.getUrl(), mIndexUrl)) {
            loadUrl();
        } else {
            binding.ewebView.onRefresh();
        }
    }

    private void restoreWebViewState(Bundle state) {
        Bundle savedState = getArguments().getBundle("savedWebViewState");
        if (savedState != null) {
            mWebView.restoreState(savedState);
        } else {
            loadUrl();
        }
    }

    private void loadUrl() {
        Bundle args = getArguments();
        String defaultUrl = Pref.getDocumentationUrl() + "index.html";
        mIndexUrl = args.getString(ARGUMENT_URL, defaultUrl);
        mWebView.loadUrl(mIndexUrl);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        Bundle webViewState = new Bundle();
        mWebView.saveState(webViewState);
        getArguments().putBundle("savedWebViewState", webViewState);
    }

    @Override
    public boolean onBackPressed(Activity activity) {
        if (mWebView.canGoBack()) {
            mWebView.goBack();
            return true;
        }
        return false;
    }

    @Override
    protected void onFabClick(FloatingActionButton fab) {
    }

    @Subscribe
    public void onQuerySummit(QueryEvent event) {
        if (!isAdded() || !isVisible()) return;

        if (event == QueryEvent.CLEAR) {
            mWebView.clearMatches();
            mPreviousQuery = null;
            return;
        }
        handleSearchQuery(event);
    }

    private void handleSearchQuery(QueryEvent event) {
        if (event.isFindForward()) {
            mWebView.findNext(false);
            return;
        }
        if (event.getQuery().equals(mPreviousQuery)) {
            mWebView.findNext(true);
            return;
        }
        mWebView.findAllAsync(event.getQuery());
        mPreviousQuery = event.getQuery();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }
}

File: app/src/main/java/org/autojs/autojs/ui/doc/DocsFragment.kt
package org.autojs.autojs.ui.doc

import android.app.Activity
import android.os.Bundle
import android.webkit.WebView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.stardust.util.BackPressedHandler
import org.autojs.autojs.Pref
import org.autojs.autojs.databinding.FragmentOnlineDocsBinding
import org.autojs.autojs.ui.main.QueryEvent
import org.autojs.autojs.ui.main.ViewPagerFragment
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

class DocsFragment : ViewPagerFragment, BackPressedHandler {

    companion object {
        const val ARGUMENT_URL = "url"
    }

    private var _binding: FragmentOnlineDocsBinding? = null
    private val binding get() = _binding!!

    private var webView: WebView? = null
    private var indexUrl: String? = null
    private var previousQuery: String? = null

    constructor() : super(ROTATION_GONE) {
        arguments = Bundle()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        EventBus.getDefault().register(this)
    }

    override fun onCreateView(inflater: android.view.LayoutInflater, container: android.view.ViewGroup?, savedInstanceState: Bundle?): android.view.View {
        _binding = FragmentOnlineDocsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: android.view.View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpViews()
    }

    private fun setUpViews() {
        webView = binding.ewebView.webView
        binding.ewebView.swipeRefreshLayout.setOnRefreshListener {
            if (webView?.url == indexUrl) {
                loadUrl()
            } else {
                binding.ewebView.onRefresh()
            }
        }
        val savedWebViewState = arguments?.getBundle("savedWebViewState")
        if (savedWebViewState != null) {
            webView?.restoreState(savedWebViewState)
        } else {
            loadUrl()
        }
    }

    private fun loadUrl() {
        indexUrl = arguments?.getString(ARGUMENT_URL) ?: (Pref.getDocumentationUrl() + "index.html")
        webView?.loadUrl(indexUrl!!)
    }

    override fun onPause() {
        super.onPause()
        val savedWebViewState = Bundle()
        webView?.saveState(savedWebViewState)
        arguments?.putBundle("savedWebViewState", savedWebViewState)
    }

    override fun onBackPressed(activity: Activity): Boolean {
        if (webView?.canGoBack() == true) {
            webView?.goBack()
            return true
        }
        return false
    }

    override fun onFabClick(fab: FloatingActionButton) {
    }

    @Subscribe
    fun onQuerySummit(event: QueryEvent) {
        if (isShown.not()) return
        if (event == QueryEvent.CLEAR) {
            webView?.clearMatches()
            previousQuery = null
            return
        }
        if (event.isFindForward) {
            webView?.findNext(false)
            return
        }
        if (event.query == previousQuery) {
            webView?.findNext(true)
            return
        }
        webView?.findAllAsync(event.query)
        previousQuery = event.query
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        webView = null
        _binding = null
    }
}
